packages/react/src/ReactLazyComponent.js

```diff
@@ -0,0 +1,26 @@
+/**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+type Thenable = {
+  then(resolve: () => mixed, reject: () => mixed): mixed,
+};
+
+export function lazy(ctor: () => Thenable) {
+  let thenable = null;
+  return {
+    then(resolve, reject) {
+      if (thenable === null) {
+        // Lazily create thenable by wrapping in an extra thenable.
+        thenable = ctor();
+      }
+      thenable.then(resolve, reject);
```

Comment:
Even though we don't use the return value, others might want to. So it might be worth returning the result of this.